src/state/slices/txHistorySlice/txHistorySlice.ts

```diff
@@ -255,48 +261,56 @@ export const txHistoryApi = createApi({
         const results: TransactionsByAccountId = {}
         await Promise.all(
           accountIds.map(async accountId => {
+            results[accountId] = []
+
             const { chainId, account: pubkey } = fromAccountId(accountId)
             const adapter = getChainAdapterManager().get(chainId)
-            if (!adapter)
-              return {
-                error: {
-                  data: `getAllTxHistory: no adapter available for chainId ${chainId}`,
-                  status: 400,
-                },
-              }
 
-            let currentCursor: string = ''
+            if (!adapter) {
+              const data = `getAllTxHistory: no adapter available for chainId ${chainId}`
+              return { error: { data, status: 400 } }
+            }
 
             try {
+              let currentCursor = ''
               do {
+                const pageSize = (() => {
+                  switch (chainId) {
+                    case gnosisChainId:
+                    case polygonChainId:
+                      return 10
+                    default:
+                      return 100
+                  }
+                })()
+
                 const { cursor, transactions } = await adapter.getTxHistory({
                   cursor: currentCursor,
                   pubkey,
-                  pageSize: 100,
+                  pageSize,
                 })
 
-                currentCursor = cursor
-
                 const state = getState() as State
-                const txState = state.txHistory.txs
-
-                const existingTxIndexes = Object.values(
-                  txState?.byAccountIdAssetId?.[accountId] ?? {},
-                ).flatMap(identity)
-
-                // freshly fetched - unchained returns latest txs first
-                const fetchedTxIndexes: TxId[] = transactions.map(tx =>
-                  serializeTxIndex(accountId, tx.txid, tx.address, tx.data),
-                )
-                // diff the two - if we haven't seen any of these txs before, upsert them
-                const diffedTxIds = difference(fetchedTxIndexes, existingTxIndexes)
-                if (diffedTxIds.length) {
-                  // new txs to return
-                  results[accountId] = transactions
-                } else {
-                  // we've previously fetched all txs for this account, don't keep paginating
-                  break
+                const txsById = state.txHistory.txs.byId
+
+                const newTxs: Transaction[] = []
+                for (const tx of transactions) {
+                  if (txsById[serializeTxIndex(accountId, tx.txid, tx.address, tx.data)]) break
+                  newTxs.push(tx)
                 }
+
+                results[accountId].push(...newTxs)
```

Comment:
mutate to add new transaction instead of previous overwrite of results on each page fetch which only added the last page fetch of transactions.